### ISSUE TITLE
offline use model

### ISSUE THREAD

**Issue by affctionP** [hexgrad/kokoro#279]
how can use model offline and not download from huggingface?

**Comment by amin-jhn** [comment]
> how can use model offline and not download from huggingface?

Hey there! I think you have to download base model once and for next runs it will automatically detect the files from cached dir.

**Comment by BajaMexico** [comment]
@amin-jhn  : this is true.  But you still have to be connected to the Internet.  I would prefer to not be connected at all.  Since its available to download model locally.

((.rolePlay_venv) ) (base) tudent12@AI-603441-13-3:~/python/rolePlay$ python3 seagull5.py
**WARNING: Defaulting repo_id to hexgrad/Kokoro-82M. Pass repo_id='hexgrad/Kokoro-82M' to suppress this warning.
'(MaxRetryError('HTTPSConnectionPool(host=\'huggingface.co\', port=443): Max retries exceeded with url: /hexgrad/Kokoro-82M/resolve/main/config.json (Caused by NameResolutionError("<urllib3.connection.HTTPSConnection object at 0x7f41bf892330>: Failed to resolve \'huggingface.co\' ([Errno -3] Temporary failure in name resolution)"))'), '(Request ID: cf2eb4c4-b4b2-40ac-b6a1-96aa5cff307e)')' thrown while requesting HEAD https://huggingface.co/hexgrad/Kokoro-82M/resolve/main/config.json
Retrying in 1s [Retry 1/5].**

**Comment by amin-jhn** [comment]
> [@amin-jhn](https://github.com/amin-jhn) : this is true. But you still have to be connected to the Internet. I would prefer to not be connected at all. Since its available to download model locally.
> 
> ((.rolePlay_venv) ) (base) tudent12@AI-603441-13-3:~/python/rolePlay$ python3 seagull5.py **WARNING: Defaulting repo_id to hexgrad/Kokoro-82M. Pass repo_id='hexgrad/Kokoro-82M' to suppress this warning. '(MaxRetryError('HTTPSConnectionPool(host='huggingface.co', port=443): Max retries exceeded with url: /hexgrad/Kokoro-82M/resolve/main/config.json (Caused by NameResolutionError("<urllib3.connection.HTTPSConnection object at 0x7f41bf892330>: Failed to resolve 'huggingface.co' ([Errno -3] Temporary failure in name resolution)"))'), '(Request ID: cf2eb4c4-b4b2-40ac-b6a1-96aa5cff307e)')' thrown while requesting HEAD https://huggingface.co/hexgrad/Kokoro-82M/resolve/main/config.json Retrying in 1s [Retry 1/5].**

You can just simply give the local path instead of hf repo
other approach is enabling HF_HUB_OFFLINE and I think it will work fine

**Comment by BajaMexico** [comment]
I tried both of those suggestions but neither worked based on code.  In the end, I downloaded the repository and modified the model.py file with the following additions (see PATCH START and PATCH_END):

`    ):
        super().__init__()
        if repo_id is None:
            repo_id = 'hexgrad/Kokoro-82M'
            print(f"WARNING: Defaulting repo_id to {repo_id}. Pass repo_id='{repo_id}' to suppress this warning.")
        self.repo_id = repo_id

        # ---- PATCH START ----
        # Load config from local path if provided
        if isinstance(config, str) and os.path.exists(config):
            with open(config, 'r', encoding='utf-8') as r:
                config = json.load(r)
        elif not isinstance(config, dict):
            if not config:
                config = hf_hub_download(repo_id=repo_id, filename='config.json')
            with open(config, 'r', encoding='utf-8') as r:
                config = json.load(r)
        # ---- PATCH END ----

        self.vocab = config['vocab']
        self.bert = CustomAlbert(AlbertConfig(vocab_size=config['n_token'], **config['plbert']))
        self.bert_encoder = torch.nn.Linear(self.bert.config.hidden_size, config['hidden_dim'])
        self.context_length = self.bert.config.max_position_embeddings
        self.predictor = ProsodyPredictor(
            style_dim=config['style_dim'], d_hid=config['hidden_dim'],
            nlayers=config['n_layer'], max_dur=config['max_dur'], dropout=config['dropout']
        )
        self.text_encoder = TextEncoder(
            channels=config['hidden_dim'], kernel_size=config['text_encoder_kernel_size'],
            depth=config['n_layer'], n_symbols=config['n_token']
        )
        self.decoder = Decoder(
            dim_in=config['hidden_dim'], style_dim=config['style_dim'],
            dim_out=config['n_mels'], disable_complex=disable_complex, **config['istftnet']
        )

        # ---- PATCH START ----
        # Load model weights from local path if provided
        if model and os.path.exists(model):
            model_path = model
        else:
            model_path = hf_hub_download(repo_id=repo_id, filename=KModel.MODEL_NAMES[repo_id])
        # ---- PATCH END ----

        for key, state_dict in torch.load(model_path, map_location='cpu', weights_only=True).items():
            assert hasattr(self, key), key
            try:
                getattr(self, key).load_state_dict(state_dict)
            except:
                state_dict = {k[7:]: v for k, v in state_dict.items()}
                getattr(self, key).load_state_dict(state_dict, strict=False)`

**Comment by FromFirstByte** [comment]
Local usage example:


<pre>
from kokoro import KPipeline
from IPython.display import display, Audio
import soundfile as sf
import os

LOCAL_GGUF = '/path/to/kokoro.gguf'

if not os.path.exists(LOCAL_GGUF):
    raise FileNotFoundError(f'Local gguf not found: {LOCAL_GGUF}')

pipeline = KPipeline(model=LOCAL_GGUF, device='mps', lang_code='a')

text = '''
[Kokoro](/kˈOkəɹO/) is an open-weight TTS model with 82 million parameters. Despite its lightweight architecture, it delivers comparable quality to larger models while being significantly faster and more cost-efficient. With Apache-licensed weights, [Kokoro](/kˈOkəɹO/) can be deployed anywhere from production environments to personal projects.
'''
generator = pipeline(text, voice='af_heart')
for i, (gs, ps, audio) in enumerate(generator):
    print(i, gs, ps)
    display(Audio(data=audio, rate=24000, autoplay=i==0))
    sf.write(f'{i}.wav', audio, 24000)
</pre>

**Comment by kenmentor** [comment]
this what i did to fix my own issue i cloned the huggi faec repo then add this main.py file 
from kokoro import KPipeline
import soundfile as sf
import os

LOCAL_CONFIG = 'config.json'
LOCAL_MODEL = 'kokoro-v1_0.pth'
VOICE_PATH = 'voices/af_heart.pt'

# Check files
if not os.path.exists(LOCAL_MODEL):
    raise FileNotFoundError(f'Local model not found: {LOCAL_MODEL}')

if not os.path.exists(LOCAL_CONFIG):
    raise FileNotFoundError(f'Local config not found: {LOCAL_CONFIG}')

if not os.path.exists(VOICE_PATH):
    raise FileNotFoundError(f'Voice file not found: {VOICE_PATH}')

# Load model manually
from kokoro.model import KModel
model = KModel(config=LOCAL_CONFIG, model=LOCAL_MODEL)

#  IMPORTANT: disable online downloads
pipeline = KPipeline(
    model=model,
    device='cpu',
    lang_code='a',
    repo_id=None   # disables huggingface
)

text = 'Hello world! Kokoro is working.'

# Load voice manually
voice = pipeline.load_voice(VOICE_PATH)

# Generate
generator = pipeline(text, voice=voice)

for i, (gs, ps, audio) in enumerate(generator):
    print(f"Segment {i}: {gs}")
    sf.write(f'{i}.wav', audio, 24000)
    print(f"Saved to `{i}.wav")`
`

**Comment by TheMasterOfDisasters** [comment]
@affctionP 
If your goal is to use Kokoro offline without downloading model/voice files from Hugging Face at runtime, my Docker-focused fork may help:

https://github.com/hangry-labs/kokoroTTS

It is designed to run Kokoro TTS as a local Docker app with:

- browser UI
- HTTP API
- prefetched model / voice assets
- offline-friendly runtime
- CPU / NVIDIA GPU examples
- WAV, MP3, FLAC, and OGG output

CPU example:

```bash
docker run -p 7860:7860 hangrylabs/kokorotts:latest
```

NVIDIA GPU example:

```bash
docker run -p 7860:7860 --gpus all hangrylabs/kokorotts:latest
```

Then open:

```text
http://localhost:7860
```

API docs are available at:

```text
http://localhost:7860/tts/docs
```

Scope note: this does not change the official `kokoro` package behavior. It is a Docker/UI/API runtime intended for users who want Kokoro running locally without manually handling Hugging Face downloads and Python dependency setup.
